### ISSUE TITLE
Long monitor contention warning

### ISSUE THREAD

**Issue by kaciula** [spotify/mobius#47]
I have this warning in my logs and I was wondering if I should be concerned. Unfortunately I could not reproduce it. What would trigger such a warning and how can I mitigate this?

`Long monitor contention with owner main (27632) at void java.lang.Object.wait!(long, int)(Object.java:-2) waiters=0 in void com.spotify.mobius.MobiusLoop$3.accept(java.lang.Object) for 109ms`

**Comment by togi** [comment]
Hello - sorry for the slow reply, I must've missed the notification.

This would indicate either that your update function is taking very long to execute, or your model dispatching is being blocked by something. If you use MobiusAndroid.controller it will by default dispatch all models to the main thread, and it would mean the problem is in your update function. Are you doing heavy calculations in there, allocating objects, or something like that? 109ms is a pretty long time for any single function to execute so there's probably a bug somewhere. Maybe run a profiler and see what it says?

**Comment by kaciula** [comment]
@togi I am using `MobiusAndroid.controller` and `RxMobius.subtypeEffectHandler()`. The update function creates new models using `copy()` based on the incoming event. Maybe the model is too big and copying takes too long. I will add some timing logging to see if I can spot the issue.

Regarding model dispatching being blocked, what exactly should I be looking for in the code?

Thanks.

**Comment by anawara** [comment]
@kaciula here's a little something that might help you figure out what's going on with your update function, if that is indeed where the bottleneck is.
```kotlin
data class InvocationTime<M, E, F>(val step: StepType<M, E, F>, val timeTaken: Long)

sealed class StepType<M, E, F>
data class Update<M, E, F>(val from: Pair<M, E>, val to: Either<Throwable, Next<M, F>>) : StepType<M, E, F>()
data class Init<M, E, F>(val from: M, val to: Either<Throwable, First<M, F>>) : StepType<M, E, F>()

sealed class Either<L, R>
data class Left<L, R>(val error: L) : Either<L, R>()
data class Right<L, R>(val value: R) : Either<L, R>()

class ProfilingLogger<M, E, F> : MobiusLoop.Logger<M, E, F> {
    private val updateMapping = mutableMapOf<Pair<M, E>, Long>()

    private var initTime = -1L
    private val trail = mutableListOf<InvocationTime<M, E, F>>()

    private inline fun time() = System.currentTimeMillis()

    override fun beforeInit(model: M) {
        initTime = time()
    }

    @Synchronized
    private fun add(pm: InvocationTime<M, E, F>) {
        trail.add(pm)
    }

    override fun afterInit(model: M, result: First<M, F>) {
        add(InvocationTime(Init(model, Right(result)), time() - initTime))
    }

    override fun exceptionDuringInit(model: M, exception: Throwable) {
        add(InvocationTime(Init(model, Left(exception)), time() - initTime))
    }

    override fun beforeUpdate(model: M, event: E) {
        updateMapping[model to event] = time()
    }

    override fun afterUpdate(model: M, event: E, result: Next<M, F>) {
        val from = model to event
        add(InvocationTime(Update(from, Right(result)), time() - updateMapping[from]!!))
    }

    override fun exceptionDuringUpdate(model: M, event: E, exception: Throwable) {
        val from = model to event
        add(InvocationTime(Update(from, Left(exception)), time() - updateMapping[from]!!))
    }

    @Synchronized
    fun inspectPerformance() = trail.toList()
}
```
You can use a `CompositeLogger` from the framework and basically use a standard logger as well as this profiling one. Then you can simply call `inspectPerformance()` to get a list of all captured invocation times.

**Comment by kaciula** [comment]
@anawara Thanks for the great code. I've profiled some screens and I sometimes get 12, 17 and even 28 ms in my update function. So I would consider this a problem. I don't do anything fancy in my code but I guess copying the model is taking too much time.

What are your typical times for the update function in your code? Do you intentionally keep your model as small as possible? Any advice on how to approach modelling?

**Comment by togi** [comment]
That sounds like a lot - normally updating models should be quite cheap, as you don't have to do a deep copy when you are using immutable objects. It is however possible to choose representations that means you have to recreate expensive objects all the time, so that is something one has to look out for. Are you looping over big data structures or anything like that? Could you maybe describe a bit more in detail what your update function is doing and what your model looks like?

**Comment by kaciula** [comment]
@togi And I profiled your sample todo app (in Kotlin) and I also do get some long durations. For instance, init in the tasks screen takes 60-62 ms and there are some other 12-15 ms durations from time to time.

@anawara Is the profiling logger bulletproof? Maybe the long durations we're seeing are not accurate.

I hope we can find the cause and have a solution for this. It's a real pleasure to work with Mobius in my apps but the performance hit is upsetting.

**Comment by kaciula** [comment]
@togi @anawara Did you get a chance to look more deeply into this issue? Thanks.

**Comment by togi** [comment]
I've been looking into it a bit but I haven't really been able to figure out a cause yet. It quickly became apparent that the way strings are created in AndroidLogger are quite expensive, but it shouldn't give you the massive readings that you are seeing. Could you try running only the profiling logger in your setup (in other words don't use CompositeLogger), and see what results you get?

Also, could you maybe try using SchedulerWorkRunners in the loop and see if that makes a difference?

**Comment by kaciula** [comment]
@togi I've run only the profiling logger and indeed, the durations are now much shorter (0 - 3 ms). So logging impacts the performance a lot. Thanks for the tip.

**Comment by pettermahlen** [comment]
I think that sounds pretty slow, still. Just out of curiosity - where are you running this? Emulator or phone? If emulator, what's the status of your hardware wrt memory pressure, etc? My hypothesis is that GC could be a culprit. To clarify that, do you see a lot of variation in the times for update executions, or are they pretty consistent around a particular duration?

**Comment by kaciula** [comment]
@pettermahlen I ran the tests both on an Samsung Galaxy S5 (Android 6) and on a genymotion emulator for Google Nexus 9.

The durations for my test app vary between 0 and 4 ms and don't seem to be very consistent.

On the other hand, the durations for the mobius todo app (for kotlin) seem to be consistent. Some are 0 ms but a few of them are 11 ms, 6 ms, 3 ms. So I would suggest you profile your todo app and maybe improve on the code if possible? My code app is heavily based on your sample app.

I'm wondering if maybe because the library is written in Java, it may not work as well with a kotlin app?

**Comment by kaciula** [comment]
@pettermahlen @anawara @togi Any update on the performance issues? I'm seeing a lot of durations of 2 - 15 ms in all the screens where I use Mobius. And the screens do feel sluggish so it's an important issue for me and my clients. I've tested on various devices ranging from Android 6 to Android 9.

Please profile your todo app in Kotlin to notice that it also suffers from this. Also, there is a small project at https://github.com/kaciula/archiman where this can also be seen.

I love coding with Mobius but the performance hit is a downer. Maybe RxMobius is to blame and I can switch to using plain Mobius. Anyway I hope there are some improvements on this front.

**Comment by togi** [comment]
Hello again! We've tried running and profiling both mobius-android-sample and your sample app, and so far the only thing we could find was a mistake in ProfilingLogger (which should use `System.nanoTime()` instead of `System.currentTimeMillis()`). Other than that, nothing seems to really take any significant time at all in the update functions, with a typical execution time of less than half of a millisecond (except if a GC pause happens, which would make it would grow to around 2-3 ms).

In your example app we could see a significant amount of time being spent in Timber (a lot of that time being to capture a stacktrace to figure out the calling method name), as well as Conductor (but only because of the UI taking time to create), but we couldn't find any traces of anything taking time in Mobius related code (once Timber was disabled).

If your apps that use mobius are behaving sluggishly even with Timber removed, i'm wondering if there is something else going on. When you say sluggishness, does that mean from time an event is sent to when the model is updated? If so we could measure that latency specifically and try to see if we can find anything. Have you run the android studio profiler on your app to see what to takes time, or have you only used the ProfilingLogger so far?

**Comment by kaciula** [comment]
@togi Thanks for looking into this. I tried using the studio profiler but only briefly. Indeed, without Timber, the app is more performant.

Do you have any suggestion on what to use for logging wihout impacting the performance?  Seeing the states, events and effects in a crash report log is extremely useful and I would not want to live without it. What do you use in production?

**Comment by togi** [comment]
There are two costly things when you log right now: Timber figuring out the calling class, and serializing the objects (currently `toString`). The first one shouldn't be a problem if you don't use DebugTree in release builds.

The second one I can think of three options: make the serialization faster, postpone serialization until a crash happens, or make logging asynchronous.

The default serialization isn't really slow, but you could make it faster by eg. only serializing the most important part of the model, or only event/effect names without individual field. The tradeoff here is that you might miss out on information that you want.

Postponing could be done by making the logger remember eg. the 50 most recent events/models/effects, and only serialize it if a crash happens. This would ofc mean you only get the information in the event of a crash and that you have to spend some extra memory to retain those 50 objects, but that should probably not be a huge deal in most cases.

One last option would be to offload serialization to a background thread so it doesn't block the loop. Loops typically don't have a very high rate of incoming events, so backpressure shouldn't be a big problem. Maybe someone could look into building such an utility?

**Comment by pettermahlen** [comment]
One such existing utility is https://logback.qos.ch/manual/appenders.html#AsyncAppender.
